# <center> Filling Missing Values in a DataFrame

In [1]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings("ignore")

In [2]:
df = pd.read_excel("./fill_na.xlsx")
df

date  temperature  windSpeed  status
0 2020-05-06      35.6582  10.788378   sunny
1 2020-05-07          NaN        NaN     NaN
2 2020-05-08      30.9343        NaN   rainy
3 2020-05-09          NaN   6.889682  cloudy
4 2020-05-10      13.9082  19.012990   rainy
5 2020-05-11      23.9382        NaN   sunny

In [3]:
df.describe().T

count       mean       std        min       25%        50%  \
temperature    4.0  26.109725  9.452383  13.908200  21.43070  27.436250   
windSpeed      3.0  12.230350  6.188951   6.889682   8.83903  10.788378   

                   75%       max  
temperature  32.115275  35.65820  
windSpeed    14.900684  19.01299

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6 entries, 0 to 5
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   date         6 non-null      datetime64[ns]
 1   temperature  4 non-null      float64       
 2   windSpeed    3 non-null      float64       
 3   status       5 non-null      object        
dtypes: datetime64[ns](1), float64(2), object(1)
memory usage: 320.0+ bytes


# <center>Filling a common value to all missing value

In [5]:
df.fillna(0) 

date  temperature  windSpeed  status
0 2020-05-06      35.6582  10.788378   sunny
1 2020-05-07       0.0000   0.000000       0
2 2020-05-08      30.9343   0.000000   rainy
3 2020-05-09       0.0000   6.889682  cloudy
4 2020-05-10      13.9082  19.012990   rainy
5 2020-05-11      23.9382   0.000000   sunny

Filled all Nan values with Zero(0)

# <center>Adding missing values to individual columns

### Filling using Dictionary 

In [6]:
df.fillna({'temperature': 30,'windSpeed': 10,'status':'sunny'})

date  temperature  windSpeed  status
0 2020-05-06      35.6582  10.788378   sunny
1 2020-05-07      30.0000  10.000000   sunny
2 2020-05-08      30.9343  10.000000   rainy
3 2020-05-09      30.0000   6.889682  cloudy
4 2020-05-10      13.9082  19.012990   rainy
5 2020-05-11      23.9382  10.000000   sunny

Filled Nan values using python dictionary in individual columns.
Column names taken as Key and datapoint as Value of Dictionary

# <center>Alternate way to fill individual columns

In [7]:
df.status.fillna('cloudy')

0     sunny
1    cloudy
2     rainy
3    cloudy
4     rainy
5     sunny
Name: status, dtype: object

# <center> Forward Fill (Row)

Forward fill is a method to forward the data from the row above the missing value. Thus all themissing value will get filled with the value above. If there are multiple missing values consecutively , they will also get filled with the same value of the above available data.

In [8]:
df.fillna(method='ffill')

date  temperature  windSpeed  status
0 2020-05-06      35.6582  10.788378   sunny
1 2020-05-07      35.6582  10.788378   sunny
2 2020-05-08      30.9343  10.788378   rainy
3 2020-05-09      30.9343   6.889682  cloudy
4 2020-05-10      13.9082  19.012990   rainy
5 2020-05-11      23.9382  19.012990   sunny

Note: The method specifies the filling method used is forward fill(ffill). The Nan value in the second row of temperature column filled by the value of the above row(35.6582)

# <center> Backward Fill (Row)

Backward fill is a method to forward the data from the below above the missing value. Thus all the missing value will get filled with the value below. If there are multiple missing values consecutively , they will also get filled with the same value of the below available data.

In [9]:
df.fillna(method='bfill')

date  temperature  windSpeed  status
0 2020-05-06      35.6582  10.788378   sunny
1 2020-05-07      30.9343   6.889682   rainy
2 2020-05-08      30.9343   6.889682   rainy
3 2020-05-09      13.9082   6.889682  cloudy
4 2020-05-10      13.9082  19.012990   rainy
5 2020-05-11      23.9382        NaN   sunny

Note: The method specifies the filling method used is backward fill(bfill). The Nan value in the second row of temperature column filled by the value of the below (30.9343)

# <center> Fordward Fill (Column)

Columnwise filling forwrd values from left side to right side.

In [10]:
df.fillna(method='ffill',axis='columns')

date          temperature            windSpeed               status
0 2020-05-06              35.6582            10.788378                sunny
1 2020-05-07  2020-05-07 00:00:00  2020-05-07 00:00:00  2020-05-07 00:00:00
2 2020-05-08              30.9343              30.9343                rainy
3 2020-05-09  2020-05-09 00:00:00             6.889682               cloudy
4 2020-05-10              13.9082             19.01299                rainy
5 2020-05-11              23.9382              23.9382                sunny

Note: Nan values of second row filled by the values from the date column.

# <center> Backward Fill (Column)

In [11]:
df.fillna(method="bfill",axis='columns')

date temperature  windSpeed  status
0 2020-05-06     35.6582  10.788378   sunny
1 2020-05-07         NaT        NaT     NaT
2 2020-05-08     30.9343      rainy   rainy
3 2020-05-09    6.889682   6.889682  cloudy
4 2020-05-10     13.9082   19.01299   rainy
5 2020-05-11     23.9382      sunny   sunny

Note: Nan values of second row (index=1) filled with NaT(Not a Time)

# <center> Limiting Forward/Backward Fill

We can limit the number of rows or columns getting filled. Use limit argument to secify the limit.

In [12]:
df.fillna(method='ffill',limit=1)

date  temperature  windSpeed  status
0 2020-05-06      35.6582  10.788378   sunny
1 2020-05-07      35.6582  10.788378   sunny
2 2020-05-08      30.9343        NaN   rainy
3 2020-05-09      30.9343   6.889682  cloudy
4 2020-05-10      13.9082  19.012990   rainy
5 2020-05-11      23.9382  19.012990   sunny

Note: Third column(wind speed) have two consecutive Nan values but here we limited to fill only 1 Nan value by giving limit=1

# <center> Filling with Pandas objects

There are many Pandas objects like df.sum(), df.max(), etc. we can fill the missing values withthese too.

In [13]:
df.fillna(df.mean())

date  temperature  windSpeed  status
0 2020-05-06    35.658200  10.788378   sunny
1 2020-05-07    26.109725  12.230350     NaN
2 2020-05-08    30.934300  12.230350   rainy
3 2020-05-09    26.109725   6.889682  cloudy
4 2020-05-10    13.908200  19.012990   rainy
5 2020-05-11    23.938200  12.230350   sunny

Note: Here all the Nan values are filled with the mean value of the DataFrame. The thing to be noted here the value of the status column is filled with Nan beacuse status is of object type.

# <center> Filling a specific range of Columns

We can do filling for a specific range of column too as:

In [14]:
df.fillna(df.mean()['temperature':'windSpeed'])

date  temperature  windSpeed  status
0 2020-05-06    35.658200  10.788378   sunny
1 2020-05-07    26.109725  12.230350     NaN
2 2020-05-08    30.934300  12.230350   rainy
3 2020-05-09    26.109725   6.889682  cloudy
4 2020-05-10    13.908200  19.012990   rainy
5 2020-05-11    23.938200  12.230350   sunny

Note: Here the Nan values of the cloumns from temperature to windspeed are filled with mean of the DataFrame.

# <center> Interpolate missing value

In short, interpolation is a process of determining the unknown values that lie in between theknown data points. We can interpolate missing values based on different methods. This is doneby an object in DataFrame as interpolate() . By default, interpolate() does linear interpolation.

## Linear interpolate

Linear interpolation involves estimating a new value by connecting two adjacent known values with a straight line.

In [15]:
df.temperature.interpolate()

0    35.65820
1    33.29625
2    30.93430
3    22.42125
4    13.90820
5    23.93820
Name: temperature, dtype: float64

Note: Here the second value of the temperature column is filled with the mean value of first and third row values of temperature [(35.65820+30.93430)/2=33.29625]

## Time interpolate

time-weighted interpolation only works on Series or DataFrames with a DatetimeIndex

#### data.interpolate(method='time')

# <center> Other Methods

Other methods includes barycentric,pchip,akima,spline,polynomial

In [16]:
df.temperature.interpolate(method='barycentric')

0    35.65820
1    39.46530
2    30.93430
3    19.32775
4    13.90820
5    23.93820
Name: temperature, dtype: float64

In [17]:
df.temperature.interpolate(method='pchip')

0    35.658200
1    34.220728
2    30.934300
3    21.496772
4    13.908200
5    23.938200
Name: temperature, dtype: float64

In [18]:
df.temperature.interpolate(method='akima')

0    35.658200
1    34.448184
2    30.934300
3    22.421250
4    13.908200
5    23.938200
Name: temperature, dtype: float64

In [19]:
df.temperature.interpolate(method='spline',order=2)

0    35.658200
1    35.076089
2    30.934300
3    20.526966
4    13.908200
5    23.938200
Name: temperature, dtype: float64

In [20]:
df.temperature.interpolate(method='polynomial',order=2)

0    35.658200
1    36.196165
2    30.934300
3    19.872606
4    13.908200
5    23.938200
Name: temperature, dtype: float64